\documentclass{article}

%\VignetteIndexEntry{Using_RTCGAToolbox}
%\VignetteDepends{RTCGAToolbox}
%\VignetteKeyword{bioinformatics}
%\VignetteKeyword{genomics}
%\VignetteKeyword{RTCGAToolbox}
%\VignettePackage{RTCGAToolbox}

\usepackage{graphicx}
\usepackage{hyperref}

\begin{document}
\SweaveOpts{concordance=TRUE}

\title{RTCGAToolbox Package}
\author{Mehmet Kemal Samur, Ph.D.\\
Department of Biostatistics and Computational Biology,\\
Dana Farber Cancer Institute and Harvard School of Public Health,\\
Boston, MA}
\date{May 25, 2014}
\maketitle

\tableofcontents

\section{Introduction}

Managing data from large scale projects such as The Cancer Genome Atlas (TCGA) for further analysis is an important and time consuming step for research projects. Several efforts, such as Firehose project, make TCGA pre-processed data publicly available via web services and data portals but it requires managing, downloading and preparing the data for following steps. We developed an open source and extensible R based data client for Firehose Level 3 and Level 4 data and demonstrated its use with sample case studies. RTCGAToolbox could improve data management for researchers who are interested with TCGA data. In addition, it can be integrated with other analysis pipelines for following data analysis.
\\

RTCGAToolbox is open-source and licensed under the GNU General Public License Version 2.0. All documentation and source code for RTCGAToolbox is freely available.
\\

Currently following functions are provided to access datasets and processed them.
\\

\begin{itemize}
\item
Control funtions:
\begin{itemize}
  \item getFirehoseRunningDates: This function can be called to access valid stddata run dates. To access data users have to provide valid dates.
  \item getFirehoseAnalyzeDates: This function can be called to access valid analyze run dates. To access data users have to provide valid dates. This function only effects the GISTIC2 processed copy estimate matrices.
  \item getFirehoseDatasets: This function can be called to access valid dataset aliases.
\end{itemize}
\item
Data client function:
\begin{itemize}
  \item getFirehoseData: This is the core function of the package. Users can access Firehose processed data via this function. Once it is called, several steps are done by the library to access data. Finally this function return an S4 object that keeps all downloaded data.
\end{itemize}
\item
Analysis Functions:
\begin{itemize}
  \item getDiffExpressedGenes: This function takes "FirehoseData" object as an input and uses differential gene expression analysis to compare cancer and normal samples. Function takes "limma" package advantages for perfroming analysis. In adddition, sample and normal population is obtained from TCGA sample barcodes.
\end{itemize}
\begin{itemize}
  \item getCNGECorrelation: This function calculates the correlation between gene expression values and copy number data. Users have to donwload GISTIC2 copy number estimates to use this function as well as the expression data from at least one platfrom.
\end{itemize}
\begin{itemize}
  \item getMutationRate: From all samples that have mutation infromation, this function get the genes mutation frequency.
\end{itemize}
\begin{itemize}
  \item getSurvival: Performs an univariate survival comparison for individual genes between high and low expressed sample groups.
\end{itemize}
\begin{itemize}
  \item getReport: Creates a circular pdf figure from differential gene expression, copy number and mutation information.
\end{itemize}
\end{itemize}

After successful installation of RTCGAToolbox, one needs to load the library to get started using it:

<<intro>>=
library(RTCGAToolbox)
@

\section{Data Client}

Before getting data from Firehose pipelines, users have to check valid dataset aliases, stddata run dates and analyze run dates. To provide valid information RTCGAToolbox provides three control functions. Users can list datasets with "getFirehoseDatasets" function. In addition, users have to provide stddata run date or/and analyze run date for client function. Valid dates can be accessible via "getFirehoseRunningDates" and "getFirehoseAnalyzeDates" functions. Following code chunk shows how to list datasets and dates.

<<Input>>=
# Valid aliases
getFirehoseDatasets()
# Valid stddata runs
stddata = getFirehoseRunningDates()
stddata
# Valid analysis running dates (will return 3 recent date)
gisticDate = getFirehoseAnalyzeDates(last=3)
gisticDate
@

Once users decide the dates that they want to use and dataset they can call data client function ("getFirehoseData") to access data. Current version can download multiple data types except ISOFORM and exon level data due to data size. ollowing code chunk will download BRCA dataset with clinical, gene expression (both from RNASeq and mRNA array platfroms), mutation and GISTIC2 procesed copy number estimates data. 

<<eval=FALSE,figs.only=TRUE>>=
# BRCA data with mRNA (Both array and RNASeq), GISTIC processed copy number data
# mutation data and clinical data 
# (Depends on bandwidth this process may take longer time)
brcaData = getFirehoseData (dataset="BRCA", runDate="20140416", gistic2_Date="20140115", 
              Clinic=TRUE, RNAseq_Gene=TRUE, mRNA_Array=TRUE, Mutation=TRUE)
@

\section{Post analyze functions}

RTCGAToolbox has analyze functions to provide basic information about datasets. Analyze function inlcudes diffrential gene expression analyze, coorelation analyze between CN and GE data, univariate survival analyze, mutation frequency table and report figure.

\subsection{Differential gene expression}

RTCGAToolbox hires voom and limma package function to preform differential gene expression analysis between "Normal" and "Cancer" tissue samples. Every samples which is prosecced by TCGA project has a structured barcode number which includes the source of the tissue. RTCGAToolbox uses the barcode information to sperate samples into "Normal" or "Tumor" groups and perform DGE analyze. Since "voom" requires raw count for RNASeq data, normalized RNASeq data can not be used for analyze.\\

This function uses all gene expression datasets and returns a list which each member is "DGEResult" object. Each result object keeps toptable fro the genes that has 2 log fold change expression difference and significant adjusted p value.

<<eval=FALSE>>=
# Differential gene expression analysis for gene level RNA data.
diffGeneExprs = getDiffExpressedGenes(dataObject=brcaData,DrawPlots=TRUE)
# Show head for expression outputs
for(i in length(diffGeneExprs))
{
  writeLines(diffGeneExprs[[i]]@Dataset)
  head(diffGeneExprs[[i]]@Toptable)
}
@

If "DrawPlots" set as TRUE, running code above will provide following image outputs.\\
\\
\includegraphics[width=\textwidth]{Figure2.png}

\subsection{Correlation between gene expression and copy number data}

"getCNGECorrelation" function provides correlation value and adjusted p value between copy number and gene expression data for each dataset. This function takes main dataobject as an input (uses gene copy number estimates from GISTIC2 algorithm and gen expression values from every platfrom (RNAseq and arrays) to prepare return lists. List object stores "CorResult" object that stores results for each comparison.)

<<CNGECor,eval=FALSE>>=
#Correlation between gene expression values and copy number
corrGECN = getCNGECorrelation(dataObject=brcaData) 

@

If the dataset has RNASeq data, data must be normalized for correlation analysis. This function does not use raw counts.

\subsection{Mutation frequencies}

"getMutationRate" function gets the data frame that stores mutation frequency for the genes. This function gets the mutation information for each sample taht has data and calculates frequency for each gene.

<<eval=FALSE>>=
# Mutation frequencies
mutFrq = getMutationRate(dataObject=brcaData)
head(mutFrq[order(mutFrq[,2],decreasing=TRUE),])

@

\subsection{Univariate survival analysis}
Survival analysis is usually considered as useful clinical information. To provide this information this function creates 2 or 3 groups based on expression data. (If the dataset has RNASeq data, data must be normalized for survival analysis. This function does not use raw counts.) If function is triggered with 2 groups, RTCGAToolbox creates groups using the median expression level of individual gene. If it is setted as 3 groups will be samples in 1st. quartile (expression < 1st Q), samples have higher expression (expression > 3rd Q) and samples between these 2 groups.

This function also needs a survival data which can be obtained using clinical data frame. Clinical data frames can be obtained from main data downlods. First column of the survival data fram must be sample barcodes, second column should be time and last column should be event data. Following code chunk shows how survival data frame can be obtained from clinical data and hpw survival analysis can be done.

<<Survival,eval=FALSE>>=
# Creating survival data frame and running analysis for  
# PIK3CA which is one of the most frequently mutated gene
clinicData = brcaData@Clinical
clinicData = clinicData[3:5,2:ncol(clinicData)]
clinicData[3,is.na(clinicData[3,])] = clinicData[2,is.na(clinicData[3,])]
survData <- data.frame(Samples=colnames(clinicData),
                       Time=as.numeric(clinicData[3,]),
                       Censor=as.numeric(clinicData[1,]))
getSurvival(dataObject=brcaData,geneSymbols=c("PIK3CA"),sampleTimeCensor=survData)

@
Running code above will provide following KM plot.\\
\\
\includegraphics[width=\textwidth]{Figure3.png}

\subsection{Report figure}
This function provides an overall figure for the dataset. This function uses differential gene expression analysis results (max results for 2 different platfrom), copy number data estimates from GISTIC2 and mutation data.\\

Outer circle shows the gene symbols that have mutation in at least 5\% of samples. Inner tracks are show the significantly altered gene expressions as fold change and copy number changes where blue represents the deletions and red represents the amplifications.\\

This function needs a gene coordina data frame which can be obtained from "hg19.ucsc.gene.locations" data object. Please see next section.

<<eval=FALSE>>=
# Creating dataset analysis summary figure with getReport. 
# Figure will be saved as PDF file.
data(hg19.ucsc.gene.locations)
getReport(dataObject=brcaData,DGEResult1=diffGeneExprs[[1]],
DGEResult2=diffGeneExprs[[2]],geneLocations=hg19.ucsc.gene.locations)
@

Running code above will provide following KM plot.\\
\\
\includegraphics[width=\textwidth]{Figure4.png}

\section{Data objects}
RTCGAToolbox provides 2 data objects.

"RTCGASample" data object provides sample data for testing functions and "hg19.ucsc.gene.locations" data object provides required gene coordinates from hg19 for report figure.\\

<<eval=FALSE>>=
data(RTCGASample)
data(hg19.ucsc.gene.locations)
@



\section{sessionInfo}

<<sessionInfo>>=
sessionInfo()
@

\newpage

\end{document}